distributions/categorical.lua

```diff
@@ -10,46 +10,46 @@ function distributions.cat.logpdf(...)
     error('Not implemented')
 end
 
-local function _idxResultAndN(x, categories)
+local function _idxResult(x, categories)
     -- return res, N
     if type(x) == 'number' then
-        return torch.LongTensor(x), x
+        return torch.LongTensor(x)
     elseif distributions._isTensor(x) then
         if categories then
             print('WARNING: cannot use result tensors and categories at the same time. Ignoring result tensor.')
             -- TODO: once index() will accept resulting tensor, use it
         end
-        return x, x:numel()
+        return x
     end
 end
 
 -- Categorical sampling
 function distributions.cat.rnd(...)
-    local I, N, p, options
+    local I, p, options
     local nArgs = select('#', ...)
     if nArgs == 1 then
         -- only p
-        I, N = _idxResultAndN(1)
         p = select(1, ...)
         options = {}
+        I = _idxResult(1)
     elseif nArgs == 2 then
         -- (N,  p) or (p, options)
         if type(select(2, ...)) == 'table' then
             -- (p, options)
             p, options = ...
             options = options or {}
-            I, N = _idxResultAndN(1, options.categories)
+            I = _idxResult(1)
```

Comment:
Yes, as `_idxResult` does not consume the second argument if the first one has type `number`.